src/Compiler/Graphs/MinCut.cs

```diff
@@ -171,7 +171,9 @@ public Set<NodeType> GetSourceGroup()
                     {
                         NodeType target = graph.TargetOf(edge);
                         bool isSinkEdge = IsSinkEdge(edge);
-                        int distTarget = isSinkEdge ? 0 : distanceToSink[target];
+
+                        // Long to avoid overflow.
+                        long distTarget = isSinkEdge ? 0 : distanceToSink[target];
```

Comment:
Instead of this, it makes more sense to include the line `if (sourceGroup.Contains(target)) continue;` as explained in the comment above.

Reply:
Added this check. And also changed the comparison from (x != y + 1) to (x - 1 != y) that removes the need to switch to long to avoid overflow.